include/cuco/detail/static_map/kernels.cuh

```diff
@@ -108,4 +108,105 @@ __global__ void insert_or_apply(InputIt first, cuco::detail::index_type n, Op op
+  cuco::detail::index_type n,
+  Op op,
+  Ref ref,
+  typename SharedMapRefType::extent_type window_extent)
+{
+  namespace cg     = cooperative_groups;
+  using Key        = typename Ref::key_type;
+  using Value      = typename Ref::mapped_type;
+  using value_type = typename std::iterator_traits<InputIt>::value_type;
+
+  auto const block       = cg::this_thread_block();
+  auto const thread_idx  = block.thread_rank();
+  auto const loop_stride = cuco::detail::grid_stride() / CGSize;
+  auto idx               = cuco::detail::global_thread_id() / CGSize;
+
+  auto warp                  = cg::tiled_partition<32>(block);
+  auto const warp_thread_idx = warp.thread_rank();
+
+  // Shared map initialization
+  __shared__ typename SharedMapRefType::window_type windows[window_extent.value()];
+  auto storage           = SharedMapRefType::storage_ref_type(window_extent, windows);
+  auto const num_windows = storage.num_windows();
+
+  __shared__ int32_t block_cardinality;
+  if (thread_idx == 0) block_cardinality = 0;
+  block.sync();
+
+  cuda::atomic_ref<int32_t, cuda::thread_scope_block> cardinality_counter(block_cardinality);
+
+  auto shared_map     = SharedMapRefType(cuco::empty_key<Key>(ref.empty_key_sentinel()),
+                                     cuco::empty_value<Value>(ref.empty_value_sentinel()),
+                                         {},
+                                         {},
+                                         {},
+                                     storage);
+  auto shared_map_ref = std::move(shared_map).with(cuco::op::insert_or_apply);
+  shared_map_ref.initialize(block);
+  block.sync();
+
+  while (idx - thread_idx < n) {
+    if constexpr (CGSize == 1) {
+      // insert-or-apply into the shared map first
+      int32_t inserted = 0;
+      if (idx < n) {
+        value_type const& insert_pair = *(first + idx);
+        inserted                      = shared_map_ref.insert_or_apply(insert_pair, op);
+      }
+
+      if (idx - warp_thread_idx < n)  // all threads in warp particpate
+      {
+        warp.sync();  // sync for inserted to materalize
+        for (int32_t i = warp.size() / 2; i > 0; i /= 2) {
+          inserted += warp.shfl_down(inserted, i);
+        }
+      }
+      if (warp_thread_idx == 0) cardinality_counter.fetch_add(inserted, cuda::memory_order_relaxed);
+      block.sync();
+      if (block_cardinality > BlockSize) break;
+    } else {
+      auto const tile = cg::tiled_partition<CGSize>(block);
+      if (idx < n) {
+        value_type const& insert_pair = *(first + idx);
+        ref.insert_or_apply(tile, insert_pair, op);
+      }
+    }
+    idx += loop_stride;
+  }
+
+  if constexpr (CGSize == 1) {
+    // write from shared_map to global_map
+    auto window_idx = thread_idx;
+    while (window_idx < num_windows) {
+      auto const slot = storage[window_idx][0];
+      if (not cuco::detail::bitwise_compare(slot.first, ref.empty_key_sentinel())) {
+        // TODO use insert ?
```

Comment:
I think we have to use `insert_or_apply` since other blocks might hold the same key. Using `insert` we would only merge one of the occurrences into the global memory map, while the other occurrences are treated as duplicates, i.e., get discarded.

Reply:
Agreed!